server/src/main/java/org/elasticsearch/search/vectors/KnnSearchBuilder.java

```diff
@@ -0,0 +1,161 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.vectors;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+import org.elasticsearch.index.query.AbstractQueryBuilder;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentFragment;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+/**
+ * Defines a kNN search to run in the search request.
+ */
+public class KnnSearchBuilder implements Writeable, ToXContentFragment {
+    private static final int NUM_CANDS_LIMIT = 10000;
+    static final ParseField FIELD_FIELD = new ParseField("field");
+    static final ParseField K_FIELD = new ParseField("k");
+    static final ParseField NUM_CANDS_FIELD = new ParseField("num_candidates");
+    static final ParseField QUERY_VECTOR_FIELD = new ParseField("query_vector");
+    static final ParseField BOOST_FIELD = AbstractQueryBuilder.BOOST_FIELD;
+
+    private static final ConstructingObjectParser<KnnSearchBuilder, Void> PARSER = new ConstructingObjectParser<>("knn", args -> {
+        @SuppressWarnings("unchecked")
+        List<Float> vector = (List<Float>) args[1];
+        float[] vectorArray = new float[vector.size()];
+        for (int i = 0; i < vector.size(); i++) {
+            vectorArray[i] = vector.get(i);
+        }
+        return new KnnSearchBuilder((String) args[0], vectorArray, (int) args[2], (int) args[3]);
+    });
+
+    static {
+        PARSER.declareString(constructorArg(), FIELD_FIELD);
```

Comment:
Did we also want to have a `filter` section under `knn` section? Or the plan is to introduce it in the follow up?

Reply:
This will be included in a follow-up. You can see the TODO list at the bottom here: https://github.com/elastic/elasticsearch/issues/87625.

Reply:
@jtibshirani How is the `filter` section for kNN different from the `query` property of search? If I understood the previous _knn_search implementation correctly they were similar?

Reply:
@sethmlarson the `query` and `knn` results will be combined together like a disjunction. This lets you combine term-based matches and vector-based matches for the same query text. The `filter` is a (non-scoring) query that limits the results that kNN search considers.

In the `_knn_search` API we put the `filter` at the top-level: https://github.com/elastic/elasticsearch/issues/81788. Here, we instead planning to nest it under the `knn` section like this:
```
"knn": {
    "field": "title_vector",
    "query_vector": [0.3f, 0.1f, ...],
    "k": 5,
    "num_candidates": 100,
    "filter": {
      "range": {
        "price": {
          "lte": 60
        }
      }
    }
  }
```

This will make it clearer that it differs from `query` and that it's specifically a prefilter on the kNN search.